flowmachine/flowmachine/core/server/query_schemas/base_exposed_query.py

```diff
@@ -0,0 +1,122 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from abc import ABCMeta, abstractmethod
+from copy import deepcopy
+
+from flowmachine.core import Query
+from flowmachine.core.query_info_lookup import QueryInfoLookup
+
+__all__ = ["BaseExposedQuery"]
+
+
+class BaseExposedQuery(metaclass=ABCMeta):
+    """
+    Base class for exposed flowmachine queries.
+
+    Note: this class and derived classes are not meant to be instantiated directly!
+    Instead, they are instantiated automatically by the class FlowmachineQuerySchema.
+
+    Example:
+
+        FlowmachineQuerySchema().load({"query_kind": "dummy_query", "dummy_param": "foobar"})
+    """
+
+    @property
+    @abstractmethod
+    def _flowmachine_query_obj(self):
+        """
+        Return the underlying flowmachine query object which this class exposes.
+
+        Returns
+        -------
+        Query
+        """
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} does not have the fm_query_obj property set."
+        )
+
+    def store_async(self):
+        """
+        Store this query using a background thread.
+
+        Returns
+        -------
+        Future
+            Future object representing the calculation.
+
+        """
+        q = self._flowmachine_query_obj
+        q.store()
+
+        # FIXME: currently we also aggregate the query here, but this is not
+        # the place to do this. Instead, there should be a separate query_kind
+        # to perform the spatial aggregation. Once we have this, the following
+        # code block should be removed and simply `q.md5` be returned as the
+        # query_id.
+        try:
+            # In addition to the actual query, also set an aggregated version running.
+            # This is the one which we return via the API so that we don't expose any
+            # individual-level data.
+            q_agg = q.aggregate()
+            q_agg.store()
+            query_id = q_agg.md5
+            # FIXME: this is a big hack to register the query info lookup also
```

Comment:
Since all `q.aggregate()` does is return `SpatialAggregate(q)`, this should be pretty easy to fix. Should do sooner rather than later, as it will change the api quite a bit.

Reply:
Agreed. Do we already have an issue for this?